akka-projection-slick/src/main/scala/akka/projection/slick/SlickProjection.scala

```diff
@@ -73,29 +74,38 @@ object SlickProjection {
         override def process(envelope: Envelope): Future[Done] = {
 
           val offset = sourceProvider.extractOffset(envelope)
-          val handlerAction = delegate.process(envelope)
-
-          sourceProvider match {
-            case vsp: VerifiableSourceProvider[Offset, Envelope] =>
-              vsp.verifyOffset(offset) match {
-                case VerificationSuccess =>
-                  // run user function and offset storage on the same transaction
-                  // any side-effect in user function is at-least-once
-                  val txDBIO = offsetStore
-                    .saveOffset(projectionId, offset)
-                    .flatMap(_ => handlerAction)
-                    .transactionally
-                  databaseConfig.db.run(txDBIO).map(_ => Done)
-                case VerificationFailure(reason) =>
-                  logger.warning(
-                    "The offset failed source provider verification after the envelope was processed. " +
-                    "The transaction will not be executed. Skipping envelope with reason: {}",
-                    reason)
-                  Future.successful(Done)
-              }
-            case _ => Future.successful(Done)
-          }
+          val txDBIO =
+            sourceProvider match {
+              case vsp: VerifiableSourceProvider[Offset, Envelope] =>
+                offsetStore
+                  .saveOffset(projectionId, offset)
+                  .flatMap(_ => delegate.process(envelope))
+                  .flatMap { action =>
+                    vsp.verifyOffset(offset) match {
+                      case VerificationSuccess => slick.dbio.DBIO.successful(action)
+                      case VerificationFailure(reason) =>
+                        logger.warning(
+                          "The offset failed source provider verification after the envelope was processed. " +
+                          "The transaction will not be executed. Skipping envelope with reason: {}",
+                          reason)
+                        slick.dbio.DBIO.failed(VerificationFailureException)
+                    }
+                  }
+                  .transactionally
+              case _ =>
+                offsetStore
+                  .saveOffset(projectionId, offset)
+                  .flatMap(_ => delegate.process(envelope))
+                  .transactionally
```

Comment:
It will be better to not repeat all the flatMap. 

For instance: 

```scala
val baseAction = 
  offsetStore
    .saveOffset(projectionId, offset)
    .flatMap(_ => delegate.process(envelope))

val finalAction = 
  sourceProvider match {
    case vsp: VerifiableSourceProvider[Offset, Envelope] => 
      baseAction.flatMap(a => doVerify(sourceProvider, a))
    case _ => baseAction
  }

finalAction.transactionally
```

btw, the verify seems to be exactly the same the one in the grouped process. So we can extract it to a private method.

Reply:
I'll re-factor the `baseAction` out, but the verification method would require a lot of type params, implicits, and parameters, so I don't think it's worth it.